File: basic_python/practices/01. Array & Hashing/ProductExceptSelf/ProdExceptSelf.cpp
#include <iostream>
#include <vector>
using namespace std;

class ProductExceptSelf {
public:

    vector<int> brute_force_1(const vector<int>& nums) {
        int n = nums.size();
        vector<int> res;

        for (int idx = 0; idx < n; idx++) {
            vector<int> arr(n);
            for (int i = 0; i < n; i++)
                arr[i] = (i != idx) ? nums[i] : 1;

            int p = 1;
            for (int x : arr) p *= x;

            res.push_back(p);
        }
        return res;
    }

    vector<int> brute_force_2(const vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n);

        for (int i = 0; i < n; i++) {
            int prod = 1;
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                prod *= nums[j];
            }
            res[i] = prod;
        }
        return res;
    }

    vector<int> using_division(const vector<int>& nums) {
        int prod = 1, zero_cnt = 0;

        for (int num : nums) {
            if (num != 0) prod *= num;
            else zero_cnt++;
        }

        int n = nums.size();
        vector<int> res(n);

        if (zero_cnt > 1)
            return vector<int>(n, 0);

        for (int i = 0; i < n; i++) {
            if (zero_cnt == 1) {
                res[i] = (nums[i] == 0 ? prod : 0);
            } else {
                res[i] = prod / nums[i];
            }
        }
        return res;
    }

    vector<int> using_presuffix_1(const vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n), pref(n), suff(n);

        pref[0] = 1;
        suff[n - 1] = 1;

        for (int i = 1; i < n; i++)
            pref[i] = nums[i - 1] * pref[i - 1];

        for (int i = n - 2; i >= 0; i--)
            suff[i] = nums[i + 1] * suff[i + 1];

        for (int i = 0; i < n; i++)
            res[i] = pref[i] * suff[i];

        return res;
    }

    vector<int> using_presuffix_2(const vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);

        int prefix = 1;
        for (int i = 0; i < n; i++) {
            res[i] = prefix;
            prefix *= nums[i];
        }

        int postfix = 1;
        for (int i = n - 1; i >= 0; i--) {
            res[i] *= postfix;
            postfix *= nums[i];
        }

        return res;
    }
};


void print_menu() {
    cout << "\nChoose a method:\n";
    cout << "1. Brute force 1\n";
    cout << "2. Brute force 2\n";
    cout << "3. Using division\n";
    cout << "4. Using prefix/suffix (method 1)\n";
    cout << "5. Using prefix/suffix (method 2)\n";
    cout << "0. Exit\n";
}


int main() {
    cout << "=== Product Except Self Solver ===\n";
    print_menu();

    int n;
    cout << "Input array length: ";
    cin >> n;

    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cout << "  Element " << i + 1 << ": ";
        cin >> nums[i];
    }

    cout << "\nYour input array: ";
    for (int x : nums) cout << x << " ";
    cout << "\n";

    ProductExceptSelf solver;

    cout << "Select option: ";
    int choice;
    cin >> choice;

    vector<int> result;

    switch (choice) {
        case 1: result = solver.brute_force_1(nums); break;
        case 2: result = solver.brute_force_2(nums); break;
        case 3: result = solver.using_division(nums); break;
        case 4: result = solver.using_presuffix_1(nums); break;
        case 5: result = solver.using_presuffix_2(nums); break;
        case 0: 
            cout << "Exiting...\n"; 
            return 0;
        default:
            cout << "Invalid option. Try again.\n";
            return 0;
    }

    cout << "Result: ";
    for (int x : result) cout << x << " ";
    cout << "\n";

    return 0;
}

File: basic_python/practices/01. Array & Hashing/ProductExceptSelf/ProdExceptSelf.py
class productExceptSelf:
    def brute_force_1(self, nums: list[int]) -> list[int]:
        res = []
        n = len(nums)

        for idx in range(n):
            arr = [nums[_] if _ != idx else 1 for _ in range(n)]
            p = 1
            for elm in arr:
                p *= elm
            res.append(p)
        return res

    def brute_force_2(self, nums: list[int]) -> list[int]:
        n = len(nums)
        res = [0] * n

        for i in range(n):
            prod = 1
            for j in range(n):
                if i == j:
                    continue
                prod *= nums[j]
            res[i] = prod
        return res

    def using_division(self, nums: list[int]) -> list[int]:
        prod, zero_cnt = 1, 0
        for num in nums:
            if num:
                prod *= num
            else:
                zero_cnt += 1

        if zero_cnt > 1:
            return [0] * len(nums)

        res = [0] * len(nums)
        for i, c in enumerate(nums):
            if zero_cnt:
                res[i] = 0 if c else prod
            else:
                res[i] = prod // c

        return res

    def using_presuffix_1(self, nums: list[int]) -> list[int]:
        n = len(nums)
        res = [0] * n
        pref = [0] * n
        suff = [0] * n

        pref[0] = suff[n - 1] = 1
        
        for i in range(1, n):
            pref[i] = nums[i - 1] * pref[i - 1]

        for i in range(n - 2, -1, -1):
            suff[i] = nums[i + 1] * suff[i + 1]

        for i in range(n):
            res[i] = pref[i] * suff[i]
        
        return res

    def using_presuffix_2(self, nums: list[int]) -> list[int]:
        res = [1] * len(nums)

        prefix = 1        
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        
        return res

def print_menu():
    print("\nChoose a method:")
    print("1. Brute force 1")
    print("2. Brute force 2")
    print("3. Using division")
    print("4. Using prefix/suffix (method 1)")
    print("5. Using prefix/suffix (method 2)")
    print("0. Exit")


if __name__ == "__main__":
    print("=== Product Except Self Solver ===")
    print_menu()

    n = int(input("Input array length: "))
    nums = [int(input(f"  Element {i+1}: ")) for i in range(n)]

    print("\nYour input array:", nums)

    solver = productExceptSelf()

    choice = input("Select option: ")

    if choice == "1":
        result = solver.brute_force_1(nums)
    elif choice == "2":
        result = solver.brute_force_2(nums)
    elif choice == "3":
        result = solver.using_division(nums)
    elif choice == "4":
        result = solver.using_presuffix_1(nums)
    elif choice == "5":
        result = solver.using_presuffix_2(nums)
    elif choice == "0":
        print("Exiting...")
    else:
        print("Invalid option. Try again.")
    print("Result:", result)